### PR TITLE
chore: change Gateway contract to be upgradable

### PR DESCRIPTION
**Pull request by skosito**
Using openzeppelin hardhat upgrades, and UUPS proxy.

Resources: https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable

UUPS vs Transparent Proxy: https://docs.openzeppelin.com/contracts/4.x/api/proxy#transparent-vs-uups

Had to go with lower package versions because ethers 5 is used. We should consider bumping ethers and other packages.

### CODE REVIEW

File: contracts/prototypes/interfaces.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.7;
+
+interface IGateway {
```
Comment (lumtis) [inline contracts/prototypes/interfaces.sol:4]: Shouldn't we add `execute` as well?

**Review by lumtis — COMMENTED** [review]
Even though we have the OZ docs, I think it would be great to add a test to show how the upgrade process would look like

**Comment by lumtis** [comment]
@coderabbitai review

File: contracts/prototypes/GatewayV2.sol [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+
+// NOTE: Purpose of this contract is to test upgrade process, the only difference should be event names
+// The Gateway contract is the endpoint to call smart contracts on external chains
+// The contract doesn't hold any funds and should never have active allowances
+contract GatewayV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
```
Comment (lumtis) [inline contracts/prototypes/GatewayV2.sol:13]: I think we could make it clear in the file name this is for test purpose like `GatewayUpgradeTest`

File: test/prototypes/GatewayUpgrade.spec.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    await token.mint(owner.address, ethers.utils.parseEther("1000"));
+
+    // Transfer some tokens to the custody contract
+    await token.transfer(custody.address, ethers.utils.parseEther("500"));
+  });
+
+  it("should upgrade and forward call to Receiver's receiveA function", async function () {
+    // Upgrade Gateway contract
+    const GatewayV2 = await ethers.getContractFactory("GatewayV2");
+    const gatewayV2 = await upgrades.upgradeProxy(gateway.address, GatewayV2);
```
Comment (lumtis) [inline test/prototypes/GatewayUpgrade.spec.ts:40]: So the `deployProxy` caller become admin and can call `upgradeProxy`?

**Review by lumtis — APPROVED** [review]
Looks good to me, we can look into have a contract upgrade workflow section in the specs

**Review by fadeev — APPROVED** [review]
Don't think this requires a review since we're not merging into `main`.

**Review by andresaiello — APPROVED** [review]

**Pull request merged**